src/kernels/kernelProvider.ts

```diff
@@ -123,25 +135,56 @@ export class KernelProvider implements IKernelProvider {
             this.disposables
         );
         this.asyncDisposables.push(kernel);
-        this.kernelsByNotebook.set(notebook, { options, kernel });
-        this.deleteMappingIfKernelIsDisposed(notebook, kernel);
+        if (notebook) {
+            this.kernelsByNotebook.set(notebook, { options, kernel });
+        } else {
+            this.kernelsByUri.set(uri.toString(), { options, kernel });
+        }
+        this.deleteMappingIfKernelIsDisposed(owner, kernel);
         return kernel;
     }
+    public getInternal(
+        owner: NotebookDocument | Uri
+    ):
+        | {
+              options: KernelOptions;
+              kernel: IKernel;
+          }
+        | undefined {
+        const notebook = workspace.notebookDocuments.find((nb) =>
```

Comment:
I believe this would skip notebooks in the interactive case?

Reply:
Ok, need a better name, lets chat in standup
We have two Uris:
* a. One is to track the original Python file for IW, or the original IPynb for Notebooks
* b. The other is the unique identifier for a Kernel
	* When we start a kernel, then we have a notebook IW or IPYnb notebook, in thie case the Uri = Uri of the Notebook Editor
	* For 3rd party extensions we don't have a notebook document, but we still need a Uri.

  
What do we call `a` & `b`
Today:
`kernel.resourceUri` (its the associated resource, in IW its *.py & in IPynb notebook its *.ipynb & when its an empty IW, then this is undefined)  
`kernel.notebookDocument` assicated with the IW notebook or IPYNB notebook

For 3rd party we need something thats unique to each kernel, & I'm proposing a Uri.
Today we have a 1-1 relationship, each kernel is associated with a notebook.
I'd like 1-1 relationship, each kernel is associated with a Uri

This way, if we want to get an existing kernel, we have a key to get it.